ecc/bls381/g1.go

```diff
@@ -466,39 +568,60 @@ func (p *G1Jac) MultiExp(curve *Curve, points []G1Affine, scalars []fr.Element)
 					index += l
 				}
 				if index != 0 {
-					buckets[index-1].AddMixed(&points[j])
+					indices[i][index-1] = append(indices[i][index-1], j)
 				}
 			}
+			chIndices[i] <- struct{}{}
+			close(chIndices[i])
+		}
+	}
+	pool.ExecuteAsyncReverse(0, int(nbCalls), work, false)
+
+	// now we have the indices, let's compute what's inside
+
+	debug.Assert(nbCalls > 1)
+	pool.ExecuteAsyncReverse(0, int(nbCalls), func(start, end int) {
+		for i := start; i < end; i++ {
+			var res G1Jac
 			sum := curve.g1Infinity
-			for j := len(buckets) - 1; j >= 0; j-- {
-				sum.Add(curve, &buckets[j])
-				tmpRes[i].Add(curve, &sum)
+			<-chIndices[i]
+			for j := len(indices[i]) - 1; j >= 0; j-- {
+				for k := 0; k < len(indices[i][j]); k++ {
+					sum.AddMixed(&points[indices[i][j][k]])
+				}
+				res.Add(curve, &sum)
 			}
+			tmpRes[i] <- res
+			close(tmpRes[i])
 		}
-	}
-	chDone := pool.ExecuteAsync(0, len(tmpRes), work, false)
+	}, false)
+
 	go func() {
-		<-chDone // that's making a "go routine" in the pool block, uncool
 		p.Set(&curve.g1Infinity)
+		debug.Assert(len(tmpRes)-2 >= 0)
 		for i := len(tmpRes) - 1; i >= 0; i-- {
 			for j := uint64(0); j < nbBits; j++ {
 				p.Double()
 			}
-			p.Add(curve, &tmpRes[i])
+			r := <-tmpRes[i]
+			p.Add(curve, &r)
 		}
 		chRes <- *p
 	}()
 	return chRes
 }
 
-// MultiExp set p = scalars[0]*points[0] + ... + scalars[n]*points[n]
-// see: https://eprint.iacr.org/2012/549.pdf
-// if maxGoRoutine is not provided, uses all available CPUs
-func (p *G1Jac) MultiExpNew(curve *Curve, points []G1Affine, scalars []fr.Element) chan G1Jac {
+type lockedG1Jac struct {
```

Comment:
is that used anywhere? @ThomasPiellard

Reply:
I am not using it, I didn't check if it is used elsewhere